src/utils/builder.ts

```diff
@@ -35,28 +35,73 @@ export function createNewPathway(name: string, description: string, pathwayId?:
   };
 }
 
+interface AddCriteriaSourceInterface {
+  updated: boolean;
+  newPathway: Pathway;
+}
+
+export function addCriteriaSourceToPathway(
+  pathway: Pathway,
+  criteria: Criteria[]
+): AddCriteriaSourceInterface {
+  let updated = false;
+  const criteriaIds = criteria.map(crit => crit.id);
+  const newPathway = produce(pathway, draftPathway => {
+    Object.entries(draftPathway.nodes).forEach(([nodeIndex, node]) => {
+      node.transitions.forEach(({ condition }, transitionIndex) => {
+        // If a matching criteria does not already exist, try and find one
+        if (condition && !criteriaIds.includes(condition.criteriaSource as string)) {
+          const [library, statement] = condition.cql.split('.');
+          const criteriaSource = criteria.find(
+            crit => crit.elm?.library.identifier.id === library && crit.statement === statement
+          )?.id;
+          // Only update if a criteria source is actually found.
+          if (criteriaSource) {
+            const condition = draftPathway.nodes[nodeIndex].transitions[transitionIndex].condition;
+            if (condition) {
+              updated = true;
+              condition.criteriaSource = criteriaSource;
+            }
+          }
+        }
+      });
+    });
+  });
+  return { updated, newPathway };
+}
+
+function addCriteriaSource(pathways: Pathway[], criteria: Criteria[]): Pathway[] {
+  const newPathways: Pathway[] = [];
+  pathways.forEach(pathway => {
+    const { newPathway } = addCriteriaSourceToPathway(pathway, criteria);
+    newPathways.push(newPathway);
+  });
+  return newPathways;
```

Comment:
This looks like a good place to use `.map` instead of `.forEach...push`.

Reply:
Good catch, much cleaner with a map